### ISSUE TITLE
[css-lists-3] Need a way to return the max value of a counter within a scope.

### ISSUE THREAD

**Issue by TUSF** [w3c/csswg-drafts#3667]
At the moment, one can define, increment and reset a counter, and then get the *current* value of that counter. However, it would be extremely useful to be able to get the maximum value of a counter within a certain scope.

For example, to mark lists or figures with "Item 2 of 6", or in the case of [Paged Media](https://drafts.csswg.org/css-page-3/), having sets of pages each with different page counters (such as for front matter, or multiple "articles" within a single document).

For this, I propose adding a `max-counter()` function, which follows the semantics of `counter()`, with the function that it outputs the *maximum* value that a counter reaches, before a counter-reset. So for example:

```CSS
h1 { counter-reset: section; }
h2 { counter-increment: section; }
h2::after { content: counter(section) " of " max-counter(section); }
```

```HTML
<h1>Title</h1>
  <h2>Section</h2>
  <h2>Section</h2>
  <h2>Section</h2>
<h1>Title</h1>
  <h2>Section</h2>
  <h2>Section</h2>
```

This should output something like:
```
Title
Section 1 of 3
Section 2 of 3
Section 3 of 3
Title
Section 1 of 2
Section 1 of 2
```

In effect, when a counter increments, they should also increment the value of every max-counter() that precedes them, except for those before a counter-reset. This function should also eliminate the need for [the immutable `pages` counter](https://www.w3.org/TR/css-page-3/#page-based-counters).

**Comment by jonjohnjohnson** [comment]
I think `counter-length()` would be a bit cleaner? At least something with `counter-` as the prefix.

**Comment by TUSF** [comment]
Good point. Would be more uniform with other counter-related properties and functions to have something with a `counter-` prefix. I'm all for `counter-length()` or `counter-max()`.

**Comment by fantasai** [comment]
I'd prefer counter-max() rather than counter-length() since if we're counting by twos, or start at 5, what's the length when we get to 10? :)

**Comment by Crissov** [comment]
`counter-final()` may be another option, while `counter-last()` or `last-counter()` feels wrong.

Such terms immediately raise the question, of course, whether there are also valid use cases for the opposite end, but I canʼt think of any useful applications for `counter-min()`, `counter-initial()` or `counter-first()`. This is probably a point in favor of `counter-length()`, perhaps `counter-size()` or `counter-range()`.

CSS counters are a special beast, otherwise `max(counter(foo))` would be an intuitive option.

**Comment by MatsPalmgren** [comment]
Note that the min/max value of a counter might be somewhere in the middle of the list.  So which functions we choose to implement really depends on what the use cases are.  E.g.
```html
<ol>
  <li>
  <li value=9>
  <li style="counter-increment: list-item -1">
</ol>
```

Here, I would say that `counter-final()` / `counter-last()` is 8, while `counter-max()` is 9, and `counter-length()` is 3.

We also need to define what these functions should return for `reversed()` counters, e.g.
```html
<ol reversed>
  <li>
  <li value=9>
  <li style="counter-increment: list-item 3">
</ol>
```
which results in the numbering: 10, 9, 12. Here, `counter-final()` is 10, `counter-last()` /  `counter-max()` is 12.

I don't think I see much use for `counter-max()` given that it can be any item.

`counter-length()` seems reasonable to implement, given that there's a use case ("step 1 of 3" etc), but then the question is which elements to count. Counting every element which has a `counter-set` or non-zero valued `counter-increment` for the counter seems reasonable. So `counter-length()` would be 3 for my examples above.
